Sources/Autolinker.php

```diff
@@ -0,0 +1,1251 @@
+		if (!self::$integrate_autolinker_schemes_done) {
+			IntegrationHook::call('integrate_autolinker_schemes', [&self::$schemes]);
+			self::$integrate_autolinker_schemes_done = true;
+		}
+
+		// For historical reasons, integrate_bbc_hook is used to give mods access to $no_autolink_tags.
+		BBCodeParser::integrateBBC();
+	}
+
+	/**
+	 * Gets a PCRE regular expression to match all known TLDs.
+	 *
+	 * @return string Regular expression to match all known TLDs.
+	 */
+	public function getTldRegex(): string
+	{
+		if (!isset($this->tld_regex)) {
+			$this->setTldRegex();
+		}
+
+		return $this->tld_regex;
+	}
+
+	/**
+	 * Gets a PCRE regular expression to match URLs.
+	 *
+	 * @return string Regular expression to match URLs.
+	 */
+	public function getUrlRegex(): string
+	{
+		if (!isset($this->url_regex)) {
+			$this->setUrlRegex();
+		}
+
+		return $this->url_regex;
+	}
+
+	/**
+	 * Gets a PCRE regular expression to match email addresses.
+	 *
+	 * @return string Regular expression to match email addresses.
+	 */
+	public function getEmailRegex(): string
+	{
+		if (!isset($this->email_regex)) {
+			$this->setEmailRegex();
+		}
+
+		return $this->email_regex;
+	}
+
+	/**
+	 * Gets a JavaScript regular expression to match all known TLDs.
+	 *
+	 * @return string Regular expression to match all known TLDs.
+	 */
+	public function getJavaScriptTldRegex(): string
+	{
+		if (!isset($this->js_tld_regex)) {
+			$this->setJavaScriptTldRegex();
+		}
+
+		return $this->js_tld_regex;
+	}
+
+	/**
+	 * Gets a series of JavaScript regular expressions to match URLs.
+	 *
+	 * @return array Regular expressions to match URLs.
+	 */
+	public function getJavaScriptUrlRegexes(): array
+	{
+		if (!isset($this->js_url_regexes)) {
+			$this->setJavaScriptUrlRegexes();
+		}
+
+		return $this->js_url_regexes;
+	}
+
+	/**
+	 * Gets a JavaScript regular expression to match email addresses.
+	 *
+	 * @return string Regular expression to match email addresses.
+	 */
+	public function getJavaScriptEmailRegex(): string
+	{
+		if (!isset($this->js_email_regex)) {
+			$this->setJavaScriptEmailRegex();
+		}
+
+		return $this->js_email_regex;
+	}
+
+	/**
+	 * Detects URLs in a string.
+	 *
+	 * Returns an array in which the keys are the start positions of any
+	 * detected URLs in the string, and the values are the URLs themselves.
+	 *
+	 * @param string $string The string to examine.
+	 * @param bool bool $plaintext_only If true, only look for plain text URLs.
+	 * @return array Positional info about any detected URLs.
+	 */
+	public function detectUrls(string $string, bool $plaintext_only = false): array
+	{
+		// An &nbsp; right after a URL can break the autolinker
+		if (strpos($string, '&nbsp;') !== false) {
+			$string = strtr($string, ['&nbsp;' => str_repeat(html_entity_decode('&nbsp;', 0, $this->encoding), 3)]);
+		}
+
+		$this->setUrlRegex();
+
+		if ($plaintext_only) {
+			$no_autolink_regex = Utils::buildRegex(self::$no_autolink_tags);
```

Comment:
should this be a static variable for reuse when parsing multiple messages?

Reply:
Good idea.